// @linaria/rollup/lib/index.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = linaria;

var _path = _interopRequireDefault(require("path"));

var _pluginutils = require("@rollup/pluginutils");

var _babelPreset = require("@linaria/babel-preset");

var _logger = require("@linaria/logger");

var _utils = require("@linaria/utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This file contains a Rollup loader for Linaria.
 * It uses the transform.ts function to generate class names from source code,
 * returns transformed code without template literals and attaches generated source maps
 */
function linaria({
  include,
  exclude,
  sourceMap,
  preprocessor,
  ...rest
} = {}) {
  const filter = (0, _pluginutils.createFilter)(include, exclude);
  const cssLookup = {};
  let config;
  const codeCache = new Map();
  const resolveCache = new Map();
  const evalCache = new Map();
  return {
    name: 'linaria',

    configResolved(resolvedConfig) {
      config = resolvedConfig;
    },

    load(id) {
      return cssLookup[id];
    },

    /* eslint-disable-next-line consistent-return */
    resolveId(importee) {
      if (importee in cssLookup) return importee;
    },

    async transform(code, id) {
      var _config, _config2;

      // Do not transform ignored and generated files
      if (!filter(id) || id in cssLookup) return;
      const log = (0, _logger.createCustomDebug)('rollup', (0, _utils.getFileIdx)(id));
      log('rollup-init', id);

      const asyncResolve = async (what, importer) => {
        const resolved = await this.resolve(what, importer);

        if (resolved) {
          log('resolve', "✅ '%s'@'%s -> %O\n%s", what, importer, resolved); // Vite adds param like `?v=667939b3` to cached modules

          return resolved.id.split('?')[0];
        }

        log('resolve', "❌ '%s'@'%s", what, importer);
        throw new Error(`Could not resolve ${what}`);
      };

      const result = await (0, _babelPreset.transform)(code, {
        filename: id,
        preprocessor,
        pluginOptions: rest
      }, asyncResolve, {}, resolveCache, codeCache, evalCache);
      if (!result.cssText) return;
      let {
        cssText
      } = result;
      const slug = (0, _babelPreset.slugify)(cssText);
      const filename = `${id.replace(/\.[jt]sx?$/, '')}_${slug}.css`;

      if (sourceMap && result.cssSourceMapText) {
        const map = Buffer.from(result.cssSourceMapText).toString('base64');
        cssText += `/*# sourceMappingURL=data:application/json;base64,${map}*/`;
      }

      cssLookup[filename] = cssText;

      if (((_config = config) === null || _config === void 0 ? void 0 : _config.command) === 'serve' && (_config2 = config) !== null && _config2 !== void 0 && _config2.root) {
        cssLookup[`/${_path.default.posix.relative(config.root, filename)}`] = cssText;
      }

      result.code += `\nimport ${JSON.stringify(filename)};\n`;
      /* eslint-disable-next-line consistent-return */

      return {
        code: result.code,
        map: result.sourceMap
      };
    }

  };
}
